rust/sbd-server/src/lib.rs

```diff
@@ -233,11 +233,78 @@ impl SbdServer {
         // read this as a prioritization of existing connections over incoming
         let connect_limit = Arc::new(tokio::sync::Semaphore::new(1024));
 
-        let weak_cslot = cslot.weak();
+        let mut bind_port_zero = Vec::new();
+        let mut bind_explicit_port = Vec::new();
+
         for bind in config.bind.iter() {
             let a: std::net::SocketAddr = bind.parse().map_err(Error::other)?;
 
-            let tcp = tokio::net::TcpListener::bind(a).await?;
+            if a.port() == 0 {
+                bind_port_zero.push(a);
+            } else {
+                bind_explicit_port.push(a);
+            }
+        }
+
+        let (mut listeners, mut l2) = tokio::join!(
+            async {
+                // bail if there are no zero port bindings
+                if bind_port_zero.is_empty() {
+                    return Vec::new();
+                }
+
+                // try twice to re-use port
+                'top: for _ in 0..2 {
+                    let mut listeners = Vec::new();
+
+                    let mut a_iter = bind_port_zero.iter();
+
+                    let a = a_iter.next().unwrap();
+                    if let Ok(tcp) = tokio::net::TcpListener::bind(a).await {
+                        let port = tcp.local_addr().unwrap().port();
+                        listeners.push(tcp);
+
+                        for a in a_iter {
+                            let mut a = *a;
+                            a.set_port(port);
+                            match tokio::net::TcpListener::bind(a).await {
+                                Err(_) => continue 'top,
+                                Ok(tcp) => listeners.push(tcp),
+                            }
+                        }
+
+                        return listeners;
+                    }
+                }
+
+                // just use whatever we can get
+                let mut listeners = Vec::new();
+                for a in bind_port_zero {
+                    if let Ok(tcp) = tokio::net::TcpListener::bind(a).await {
+                        listeners.push(tcp);
+                    }
+                }
+                listeners
+            },
+            async {
+                let mut listeners = Vec::new();
```

Comment:
Could extract function and share with  https://github.com/holochain/sbd/pull/22/files#diff-705e63f98cd04e33cfe76ed073288fef626c8a97b4b8877d6eb0a6f5b2741ebdR281-R287

Not too important though!

Reply:
haha, true!

Reply:
done: https://github.com/holochain/sbd/pull/22/commits/f54f03dfe83b3e85493020e6a68d2f81feb54a6b